enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,235 @@
+deployment even started, we want to reduce wasted time and resources, customer
+escalations, and improve the perception of OpenShift deployments on-premise.
+
+The installer already performs some validations:
+- checks all required fields are set and that the data is in the right format
+- some basic validation, such as networks do not overlap
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's endpoints
+- necessary cloud services are available
+- storage performance
+
+### Non-Goals
+
+Implementing every useful validation we can think of is out-of-scope:
+- The potential number of validations is huge
+- We expect more validations to be added over time as new issues are
+  discovered.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories
+
+#### On-premise deployments
+
+As an administrator of an OpenShift cluster, I would like to verify that my
+OpenStack cloud meets all the performance, service and networking requirements
+necessary for a successful deployment.
+
+#### CI debugging
+
+As an OpenShift developer, I would like to know if the CI flakes might be
+caused by transient environmental issues.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The validations must leave the environment unaltered.
+
+The framework should allow implementing checks common to all platforms as well
+as per-platform checks.
+
+We do not envision the need for the users to write additional validations. As
+a consequence, the validations do not need to be loaded on startup and will be
+compiled into the `openshift-install` binary. This may be revisited later.
+
+#### Enabling the validations
+
+The validations will be split into two groups: the core validations (including
+all current validations) and the extra validations.
+
+Core validations are the ones we already know. They run every time, as it is
+done today.
+
+The extra validations will be enabled on demand via a flag when running the
+installer.
+
+In the future we may also consider adding in the form of a separate command or
+flag a way to perform all validations and check that the environment is
+suitable without actually performing a deployment.
+
+#### Reporting errors
+
+A failed validation typically causes the installer to fail early and not
+proceed with the deployment of OpenShift. While this is fine when the
+validation identifies a missing hard requirement, there are cases where the
+environment doesn't match the recommendation and we may want the installer to
+go on with the deployment still.
+
+In that case, the validation can be marked as optional, meaning failure of the
```

Comment:
How do you intend  to set up the system to mark tests optional? Command line flags? A YAML file?